### ISSUE TITLE
Support for https

### ISSUE THREAD

**Issue by ZuSe** [holic/redirect.name#23]
I love your service!
Great idea and damn simple to use.

Is there any chance that https is supported as well?
As I see there is right now no https support enabled for the sld alias.redirect.name
Maybe it could be setup by using acme v2 wildcard certs from letsencrypt?

**Comment by holic** [comment]
Thanks for the kind words!

There's no HTTPS support right now, but I'd be happy to have it. Do you want to open a PR?

**Comment by emaildano** [comment]
@holic Would HTTPs even be possible? I love this simple approach and have a use case for it. If HTTPs was possible I'd put some time toward exploring this.

I'm thinking there may be a conflict with the HTTPs Certs. Not just a code thing.

**Comment by Vadorequest** [comment]
Very much interested with https support, but I don't understand why it's not supported by default. Why does it need a special configuration? The redirect from one domain to another should work regardless to each certificates, no?

**Comment by Fonz001** [comment]
For those interested, it's not as trivial as at seems. To perform an HTTPS redirect without an SSL certificate is impossible (https://stackoverflow.com/questions/6331371/how-to-redirect-https-to-http-without-any-ssl-certificate). It would imply that redirect.name should integrate a service like LetsEncrypt and this also requires (a lot of) extra DNS settings.

**Comment by schlomo** [comment]
@holic Is there a reason to not integrate #26?

**Comment by mrexodia** [comment]
Recently I discovered that you can use CloudFlare with the 'Flexible' encryption mode to support https with `redirect.name`, might be worth looking into!

**Comment by mrexodia** [comment]
Ended up implementing a generic CloudFlare worker that uses the `URI` record to figure out where to redirect to https://github.com/mrexodia/cloudflare-redirect

**Comment by 9072997** [comment]
I wouldn't mind putting in work on this, but I would like to know what the requirements would be. I'm sort-of continuing a discussion from #26 .

Persistent storage is definitely the "normal" way to do this. If we use the existing #26, I think we should implement a [DecisionFunc](https://pkg.go.dev/github.com/caddyserver/certmagic#OnDemandConfig) to make sure the domain is pointed to redirect.name. Also, it looks like that PR is using the default filesystem storage ($HOME/.local/share/certmagic). That is probably a good choice, but I don't know the hosting environment, so I can't say for sure. I'd be happy to work with any form of persistent storage you have available (SQL, KV-store, filesystem).

If persistent storage is absolutely out of the question, there is still a way I can think of to still make this work. We could generate deterministic private keys based on the domain name and a secret [pepper](https://en.wikipedia.org/wiki/Pepper_(cryptography)), then obtain certificates as usual. We would cache keys/certificates in memory (or ephemeral local storage). In the event of a "cache miss" (process restart or clearing of local storage) the key could be re-constructed and the certificate fetched from a [CT log](https://letsencrypt.org/docs/ct-logs/). I don't like this solution because deterministic private keys are easy to get wrong, and I don't know what kind of performance issues we would run into trying to fetch certificates from CT-logs on-demand.

If there is no interest in either solution, I might look at maintaining and hosting a fork with HTTPS support.

**Comment by 9072997** [comment]
[I am hosting a version of this with HTTPs support](https://jons.tools/redirect/)

**Comment by frolic** [comment]
Awesome to see this! Thanks for taking the lead.

I'm fine with storing certs on disk since they're relatively lightweight. But I'd like to preserve as much "statelessness" as possible. If the disk was wiped, what is the impact on serving HTTPS requests? Would caddy/certmagic just generate a new certificate in place of the missing one? Are there any DNS propagation concerns with this?

**Comment by 9072997** [comment]
> If the disk was wiped, what is the impact on serving HTTPS requests? Would caddy/certmagic just generate a new certificate in place of the missing one?

Correct. A new certificate would be requested the next time a client tried to visit that particular domain. This *might* create an issue with Let's Encrypt rate limits (specifically the "[New Orders per Account](https://letsencrypt.org/docs/rate-limits/#new-orders-per-account)" limit of 300/3hrs). If you think of the certificate store as a sort of cache, most requests are cache hits. After a disk wipe though, every unique domain is a cache miss and we can only have 300 cache misses in 3 hours.

Possible mitigations for this include
- Just deal with it: Disk wipes are probably infrequent and even if one happened, eventually you would get back to a steady state where you had certificates for all the domains receiving regular traffic.
- [Request an override](https://isrg.formstack.com/forms/rate_limit_adjustment_request): Let's Encrypt offers overrides for some rate limits, and this is one of them. In order to make this work the ACME account would need to remain stable across disk-wipes. As things are currently configured in my fork, the account registration is handled automatically and is therefore part of the "state" that would get lost on a disk wipe. If this is the route you would like to pursue, let me know and I would be happy to make changes so that the appropriate account secrets could be be passed in as environment variables.
- Use ZeroSSL for certificates: [ZeroSSL appears to have less stringent rate limits](https://help.zerossl.com/hc/en-us/articles/17864245480093-Advantages-over-Using-Let-s-Encrypt). Preliminarily it looks like you can do unlimited certs on the free tier via ACME, and the up-sell to paid plans is mostly about the REST API and 1 year certs, but I would have to do more research to confirm this. If this is a route you are interested in, let me know.

> Are there any DNS propagation concerns with this?

Specifically related to a disk wipe? No.
Broadly? Mostly no...
Validation is performed via the [HTTP-01 challenge](https://letsencrypt.org/docs/challenge-types/#http-01-challenge), so the only DNS change we have to worry about is the one the user makes to point their domain to the IP this application is running on. If we try to request a cert before that propagates, validation fails. It's not that bad if this happens, but it does count against the "New Orders per Account" limit. I added a check to see if the domain appears to be correctly configured *from our point of view*, but because of DNS propagation delays it is possible it doesn't yet appear correctly configured *from Let's Encrypt's point of view*. In practice, Let's Encrypt is highly optimized for this. In my experience, Let's Encrypt always picks up DNS changes before my local resolver.

**Comment by schlomo** [comment]
Maybe stupid question: What's about about backing up the state e.g. every hour or X minutes? That way the only thing lost is the "new" state that came in since the last backup and you probably won't hit the rate limit.

I suppose that the state is not very large, so that even a periodic upload to an object store won't cost very much.

**Comment by frolic** [comment]
There's tons of options to preserve state, of course. I am just trying to understand the worst case scenario to decide how much state I actually want to take responsibility for. I don't collect logs so I have very little insight into how much traffic this service is doing and how many unique domains there are, but the rate limit is a good enough reason to attempt to do some sort of disk caching/backups.

**Comment by schlomo** [comment]
TBH, I'd expect people who use an anonymous redirect service to also accept that they can do whatever they want with this effective grant into their own domain. Your storing the state and certificates adds IMHO very very little actual risk to the game.

Maybe you can simply add a stronrg-worded disclaimer like this:

> By using the redirect.name service I acknowledge that the operator of redirect.name gains full access to the content hosted under the DNS name that I point to it, including generating valid TLS certificates for that name and potentially serving other - potentially hostile - content there. There is no warranty or liability accepted for any harm done to me from my usage of redirect.name and I'm aware that I should be hosting my own instance if I truly care about security and privacy for the users of the DNS name I point to redirect.name.

I mean, that is what it actually is and it totally kind of you to run the service and offer it for free.
